node/src/components/block_proposer.rs

```diff
@@ -402,32 +344,51 @@ where
                     .set_timeout(PRUNE_INTERVAL)
                     .event(|_| Event::BufferPrune);
             }
-            Event::Request(BlockProposerRequest::ListForInclusion {
-                current_instant,
-                past_blocks,
-                responder,
-            }) => {
-                return self.get_chainspec(effect_builder, current_instant, past_blocks, responder);
+            Event::Request(BlockProposerRequest::ListForInclusion(request)) => {
+                if request.next_block_height > self.state.next_finalized {
+                    self.state
+                        .request_queue
+                        .entry(request.next_block_height)
+                        .or_default()
+                        .push(request);
+                } else {
+                    return self.get_chainspec(effect_builder, request);
```

Comment:
This will get executed also if `request.next_block_height < self.state.next_finalized`. This would be a bug though, right? Should we panic instead?

Reply:
Would it really be a bug? If we can't rely on event ordering, the request may be sent before the finalized block but handled after it.

Reply:
`self.state.next_finalized` seems to be updated only when it processes the new finalized block, that has all of its dependencies (including finalized parent satisfied). If we now see finalization of a block with a lower block height it means that we had "refinalized" a block.

Reply:
But couldn't it happen in theory that:
* We get a chance to propose block 5, so consensus fires a request for height 5, which gets delayed.
* In the meantime, consensus receives a competing proposal for block 5, and it gets finalized, and then the same for block 6.
* The finality announcements arrive and the block proposer updates `next_finalized` to 7.
* The request arrives, asking for proposals on height 5. (I guess at that point we could ignore the request?)

Also, isn't `request.next_block_height == self.state.next_finalized` totally fine, if the parent of the block we're about to propose has already been finalized?

Reply:
Right, I didn't think about competing blocks at the same height. You're correct, we should probably ignore the request instead of panicking.

Reply:
I think that's okay. It's not a problem if some more blocks get finalized between the point when the request was made and the point when we handle it. It won't cause us to include any deploys that have already been included in a block, and that's the main purpose here.

Reply:
But it would be wasteful to handle a block proposal at some height if we had already finalized competing block at that height, right?

Reply:
> The request arrives, asking for proposals on height 5.

But the height here isn't the target height for the proposal (see, that's the confusing part I've mentioned on Slack! :stuck_out_tongue: ) - it's just the next finalized block consensus has been expecting when it made the request. It's a way of expressing which blocks have been finalized at that point, not what the target height is (we could just as well be making a request for height 8 when next block to be finalized was 6, because 6 and 7 have already been proposed, just not finalized yet).

So I think we should be handling the request normally, not ignore it and not panic.

Reply:
> But it would be wasteful to handle a block proposal at some height if we had already finalized competing block at that height, right?

See above - it's not about the target height for the proposal. We can still be expected to propose the next block even if some blocks have been finalized in the meantime. I don't think it's wasteful.

Reply:
I understand that requests for proposal/finalization may arrive out of order. My concern is about a situation where we had already finalized block at height=10 (and did it properly, updating `self.state.next_finalized`) and suddenly we receive a notification about `ListForInclusion { height = 5 }`. It would be a bug in the consensus but it would have been handled here.

Reply:
Ah, you're right! :facepalm: 
But in my example, it would still be `<` here, would it? So at least we shouldn't panic.
(And maybe `next_block_height` should also be called `next_finalized`?)

Reply:
> My concern is about a situation where we had already finalized block at height=10 (and did it properly, updating self.state.next_finalized) and suddenly we receive a notification about ListForInclusion { height = 5 }. It would be a bug in the consensus but it would have been handled here.

I don't think it would be a bug, I think it could theoretically legitimately happen. Say the scenario is like this:

1. We want to propose a block. Our last finalized block is at height 4, but blocks 5, 6, 7, 8, 9 have been proposed already, so we're proposing block 10 now. We fire a request: `ListForInclusion { next_block_height = 5 }` (last finalized + 1)
2. Blocks 5-9 become finalized, notifications are fired.
3. Due to variations in ordering, Block Proposer receives the finality notifications and handles them, thus bumping `next_finalized` to 10.
4. Block Proposer receives the `ListForInclusion { height = 5 }`.

The only "problem" here is that the `ListForInclusion` request will (should) contain deploys from blocks 5-9 in `past_deploys`, and now they are also in `finalized_deploys` - but that's a non-issue, as we will exclude them from our proposal all the same. So I think we should handle the request normally.

Reply:
> (And maybe next_block_height should also be called next_finalized?)

Yes, that may be a good idea. :+1:

Reply:
OK, I am already confused with who fires what and when :dizzy_face:  . If you're convinced this won't be a problem then I choose to believe you.

Reply:
I tried explaining this a bit more in the comments now. Maybe this will help clarify things :grimacing: